File: tools/capture_phc06_land_hearing_authority_copy.py
#!/usr/bin/env python3
"""PHC-06 evidence: render the Land "Where this stands" authority panel
(site/land_authority_summary_view.mjs's landAuthoritySummaryHTML(), which now
includes the plain-role line from site/land_hearing_authority_copy.mjs) and a
literal reproduction of the compressed upcoming-hearing row
(site/app/land.mjs's landHearingRowHTML(), not exported — reproduced verbatim
by tools/capture_phc06_land_hearing_authority_copy.mjs and checked against the
real source below before capture) for the specimens named in the card, at both
review widths. Runs axe-core against each rendered fragment.

No image is written or committed — see docs/evidence/public-input-consequence/
phc-06-land-use-authority-in-plain-terms/capture-manifest.json for the
committed proof (content hash + textual assertion per case), per the
workstream's no-committed-image-binaries convention (spec.md).

  node tools/capture_phc06_land_hearing_authority_copy.mjs   # (invoked by this script)
  python3 tools/capture_phc06_land_hearing_authority_copy.py
"""

from __future__ import annotations

import hashlib
import json
import shutil
import subprocess
import sys
import tempfile
import time
from pathlib import Path

sys.path.insert(0, str(Path(__file__).parent.parent / "test/functional/assets"))
from a11y_gate import failing_violations  # noqa: E402

from playwright.sync_api import sync_playwright

ROOT = Path(__file__).resolve().parents[1]
AXE = ROOT / "test/functional/assets/axe.min.js"
VIEWPORTS = (("mobile", 390, 844), ("desktop", 1440, 900))
TMP_DIR = ROOT / "site/.phc06-capture-tmp"
OUT_DIR = ROOT / "docs/evidence/public-input-consequence/phc-06-land-use-authority-in-plain-terms"

# The upcoming-hearing row template the .mjs helper reproduces is not exported
# from site/app/land.mjs — these substrings must still be present in the real
# source, so a later edit to the real row markup fails this capture instead of
# silently drifting from the literal reproduction it captures against.
ROW_FIDELITY_MARKERS = (
    'import { landHearingRowRoleHTML } from "../land_hearing_authority_copy.mjs"',
    "landHearingRowRoleHTML(landAuthoritySummaryFor(row)",
    '${roleTxt?` · ${roleTxt}`:""}',
)


def check_row_template_fidelity() -> None:
    source = (ROOT / "site/app/land.mjs").read_text(encoding="utf-8")
    missing = [marker for marker in ROW_FIDELITY_MARKERS if marker not in source]
    if missing:
        raise SystemExit(
            "site/app/land.mjs no longer matches the literal row template this capture "
            f"reproduces; missing: {missing}"
        )


def start_local_site_server() -> tuple[subprocess.Popen, str]:
    ready_file = Path(tempfile.mktemp(prefix="crol-phc06-site-"))
    proc = subprocess.Popen(
        [sys.executable, str(ROOT / "tools/local_site_server.py"),
         "--directory", "site", "--port", "0", "--ready-file", str(ready_file)],
        cwd=ROOT,
    )
    deadline = time.time() + 20
    while time.time() < deadline:
        if ready_file.exists() and ready_file.stat().st_size:
            base = ready_file.read_text().strip()
            if base:
                return proc, base
        if proc.poll() is not None:
            raise RuntimeError("local_site_server.py exited before becoming ready")
        time.sleep(0.1)
    raise RuntimeError("timed out waiting for local_site_server.py")


def panel_css() -> str:
    html = (ROOT / "site" / "index.html").read_text(encoding="utf-8")
    import re

    rules = re.findall(
        r"(\.land-authority-[^{]+\{[^}]+\}|\.land-hearing-row[^{]*\{[^}]+\}|\.row \.[a-z]+\{[^}]+\}|"
        r"\.rtitle\{[^}]+\}|\.rmeta\{[^}]+\}|\.ragency\{[^}]+\}|\.fcard-compact-actions[^{]*\{[^}]+\}|"
        r"\.act[,{][^{]*\{[^}]+\})",
        html,
    )
    root = (
        ":root{--rule:#d6d3cd;--oxblood:#7a1f2b;--muted:#5c5852;--ink:#1c1917;--blue:#1a44e0;}"
        "body{margin:16px;background:#f4f1ea;font:16px/1.4 ui-sans-serif,system-ui,sans-serif;}"
        "a{color:#174ea6}"
    )
    return root + "".join(rules)


def main() -> None:
    check_row_template_fidelity()
    if TMP_DIR.exists():
        shutil.rmtree(TMP_DIR)
    generated = subprocess.run(
        [shutil.which("node") or "node", str(ROOT / "tools/capture_phc06_land_hearing_authority_copy.mjs")],
        cwd=ROOT, capture_output=True, text=True, check=True,
    )
    cases = json.loads(generated.stdout)

    css = panel_css()
    (TMP_DIR / "capture.css").write_text(css, encoding="utf-8")

    server_proc, base = start_local_site_server()
    base = base.rstrip("/")
    captures = []
    try:
        with sync_playwright() as p:
            browser = p.chromium.launch()
            for viewport_name, width, height in VIEWPORTS:
                page = browser.new_page(viewport={"width": width, "height": height})
                for case in cases:
                    fragment_url = f"{base}{case['path']}"
                    fragment_html = page.request.get(fragment_url).text()
                    # Wrapped in the same <main id="main"> landmark the real Land page
                    # renders this panel/row inside, so axe's landmark/region rules
                    # reflect the real page rather than a bare-fragment capture artifact.
                    wrapped = (
                        "<!doctype html><html lang='en'><head><meta charset='utf-8'>"
                        f"<style>{css}</style></head><body><main id='main'>{fragment_html}</main></body></html>"
                    )
                    page.set_content(wrapped, wait_until="domcontentloaded")
                    main_html = page.locator("main#main").inner_html()
                    content_sha256 = hashlib.sha256(main_html.encode("utf-8")).hexdigest()

                    if case["kind"] == "panel":
                        role_locator = page.locator("[data-land-authority-plain-role]")
                    else:
                        role_locator = page.locator("[data-land-hearing-row-role]")
                    role_present = role_locator.count() > 0
                    role_text = role_locator.first.inner_text() if role_present else None

                    page.add_script_tag(path=str(AXE))
                    result = page.evaluate(
                        "async () => await axe.run(document.querySelector('main#main') || document, "
                        "{resultTypes:['violations']})"
                    )
                    wcag22_rules = set(page.evaluate("() => axe.getRules(['wcag22aa']).map(r => r.ruleId)"))
                    gate = failing_violations(result["violations"], wcag22_rules)

                    captures.append({
                        "fixture": case["id"],
                        "kind": case["kind"],
                        "route": (
                            f"landAuthoritySummaryHTML(summary) — project:{case['project_id']}"
                            if case["kind"] == "panel"
                            else f"landHearingRowHTML(row) [literal reproduction, fidelity-checked] — project:{case['project_id']}"
                        ),
                        "viewport": viewport_name,
                        "assertion": case["assertion"],
                        "content_sha256": content_sha256,
                        "plain_role_present": role_present,
                        "plain_role_text": role_text,
                        "axe_critical_or_serious_violations": [v["id"] for v in gate],
                        "file": None,
                    })
                    print(f"{viewport_name:8s} {case['id']:40s} role={role_text!s:35} "
                          f"sha256={content_sha256[:12]} violations={[v['id'] for v in gate]}")
                page.close()
            browser.close()
    finally:
        server_proc.terminate()
        try:
            server_proc.wait(timeout=5)
        except subprocess.TimeoutExpired:
            server_proc.kill()
        shutil.rmtree(TMP_DIR, ignore_errors=True)

    repository_revision = subprocess.run(
        ["git", "rev-parse", "HEAD"], cwd=ROOT, capture_output=True, text=True, check=True,
    ).stdout.strip()

    OUT_DIR.mkdir(parents=True, exist_ok=True)
    manifest = {
        "schema_version": 1,
        "card": "cityscroll-public-input-consequence/phc-06-land-use-authority-in-plain-terms",
        "capture_mode": "local_node_playwright_axe_static_render",
        "repository_revision": repository_revision,
        "note": (
            "The authority-panel cases render landAuthoritySummaryHTML() directly against "
            "committed land-project fixtures (site/data/*.json, "
            "test/fixtures/land_authority_summary/elurp_197e_corpus.v1.json) via node, with no "
            "live Worker/D1 backend. The hearing-row cases render the new "
            "landHearingRowRoleHTML() projection composed with a literal reproduction of "
            "site/app/land.mjs's landHearingRowHTML() markup (that function is not exported); "
            "tools/capture_phc06_land_hearing_authority_copy.py checks fixed substrings of the "
            "real source before capture so a later edit to the real row markup fails this "
            "capture rather than silently drifting from it. Fragments are served from site/ "
            "(real /index.html CSS rules) via tools/local_site_server.py. axe-core "
            "(test/functional/assets/axe.min.js) ran against each rendered fragment at both "
            "review widths, using the same failing-violation classification as "
            "test/functional/11_accessibility.py (a11y_gate.py). Evidence proves the new "
            "plain-role line/span renders for a resolved current_role and does not render for an "
            "unresolved one (A1-A6), and introduces no new axe violation; it does not evidence a "
            "live-data-populated page, since no Worker/D1 backend was running."
        ),
        "data_vintage": "committed_fixture_snapshots_as_checked_in",
        "viewports": [
            {"name": "mobile", "width": 390, "height": 844},
            {"name": "desktop", "width": 1440, "height": 900},
        ],
        "captures": captures,
        "accessibility": {
            "tool": "axe-core (test/functional/assets/axe.min.js)",
            "scope": "main#main",
            "viewports_checked": ["390x844", "1440x900"],
            "new_violations_introduced": sum(len(c["axe_critical_or_serious_violations"]) for c in captures),
            "pre_existing_violations_observed": [],
        },
    }
    (OUT_DIR / "capture-manifest.json").write_text(
        json.dumps(manifest, indent=2) + "\n", encoding="utf-8"
    )
    print(f"wrote {OUT_DIR / 'capture-manifest.json'}")


if __name__ == "__main__":
    main()

File: tools/capture_phc06_land_hearing_authority_copy.mjs
#!/usr/bin/env node
/**
 * PHC-06 evidence helper: render the Land "Where this stands" authority panel
 * (site/land_authority_summary_view.mjs's landAuthoritySummaryHTML(), which
 * now includes the new plain-role line from
 * site/land_hearing_authority_copy.mjs) and a literal reproduction of the
 * upcoming-hearing row (site/app/land.mjs's landHearingRowHTML(), which is
 * not exported — its structure is reproduced here verbatim and checked
 * against the real source by tools/capture_phc06_land_hearing_authority_copy.py
 * before capture) for the specimens named in the card: a decision-maker with
 * prior advisory recommendations, a conditional decision-maker, an
 * advisory-only body, a pre-application/certification administrative
 * certifier, an expedited (ELURP) terminal decision-maker, and an unresolved
 * record with no current role. Writes each rendered fragment under
 * site/.phc06-capture-tmp/ so the .py companion can serve it locally (real
 * /index.html CSS rules) and run axe-core against it. Prints the case
 * manifest (id, path, assertion, kind) as JSON to stdout.
 *
 *   node tools/capture_phc06_land_hearing_authority_copy.mjs
 */
import { mkdirSync, readFileSync, writeFileSync } from "node:fs";
import { createRequire } from "node:module";
import { fileURLToPath } from "node:url";
import path from "node:path";

import { buildLandAuthoritySummary } from "../site/land_authority_summary.mjs";
import { landAuthoritySummaryHTML } from "../site/land_authority_summary_view.mjs";
import { landHearingRowRoleHTML } from "../site/land_hearing_authority_copy.mjs";

const ROOT = path.dirname(path.dirname(fileURLToPath(import.meta.url)));
const OUT_DIR = path.join(ROOT, "site/.phc06-capture-tmp");
const requireJson = createRequire(import.meta.url);

const geography = requireJson(path.join(ROOT, "site/data/community_board_geography_lookup.json"));
const warehouse = requireJson(path.join(ROOT, "site/data/zap_projects_warehouse_lookup.json"));
const landDefault = requireJson(path.join(ROOT, "site/data/land_default_ulurp.json"));
const hearings = requireJson(path.join(ROOT, "site/data/land_upcoming_hearings.json"));
const elurpCorpus = requireJson(path.join(ROOT, "test/fixtures/land_authority_summary/elurp_197e_corpus.v1.json"));

const i18nSrc = readFileSync(path.join(ROOT, "site/i18n.js"), "utf8");
const copy = {};
for (const match of i18nSrc.matchAll(/^\s+([a-z0-9_]+):\s*"((?:\\.|[^"\\])*)"/gm)) {
  copy[match[1]] = match[2].replace(/\\"/g, '"');
}
function t(key, vars) {
  let text = copy[key] || key;
  if (vars) {
    text = text.replace(/\{(\w+)\}/g, (all, name) => (Object.hasOwn(vars, name) ? String(vars[name] ?? "") : all));
  }
  return text;
}
function esc(value) {
  return String(value ?? "").replace(/[&<>"]/g, (ch) => ({ "&": "&amp;", "<": "&lt;", ">": "&gt;", '"': "&quot;" }[ch]));
}

function warehouseRow(projectId) {
  return (warehouse.rows || []).find((row) => row.project_id === projectId);
}
function defaultRow(projectId) {
  return (landDefault.projects || []).find((row) => row.project_id === projectId);
}
function projectHearings(projectId) {
  return (hearings.hearings || []).filter((row) => row.project_id === projectId);
}
function summarize(projectId) {
  const project = defaultRow(projectId) || warehouseRow(projectId);
  const outcome = landDefault.outcomes?.by_project?.[projectId] || {};
  return buildLandAuthoritySummary({
    project,
    geography,
    outcomes: { dispositions: outcome.dispositions, generated_at: outcome.generated_at || landDefault.generated_at },
    publishedOpportunities: { hearings: projectHearings(projectId), generated_at: hearings.generated_at },
    asOf: landDefault.generated_at,
    generatedAt: landDefault.generated_at,
  });
}
function summarizeElurp(specimenKey) {
  const specimen = elurpCorpus.specimens[specimenKey];
  const project = requireJson(path.join(ROOT, `test/fixtures/land_phase_spine/${specimen.project_id}.json`));
  return buildLandAuthoritySummary({ ...project, geography, asOf: elurpCorpus.as_of, generatedAt: elurpCorpus.as_of });
}

const ADVISORY_ONLY_SUMMARY = {
  schema: "cityscroll.land_authority_summary.v1",
  status: "resolved",
  project_id: "PHC06-ADVISORY-ONLY",
  procedure_id: "ulurp_197c",
  current_stage: { stage_id: "ulurp_197c.community_board_review", spine_phase_id: "community_board", status: "known" },
  current_actor_refs: ["community-board:brooklyn-cb-11"],
  current_role: "advisory_reviewer",
  effect: "May hold a public hearing and submit a written recommendation; the recommendation is advisory, not binding.",
  source_basis: {
    profile: { source_type: "reviewed_static_registry", procedure_id: "ulurp_197c", stage_id: "ulurp_197c.community_board_review", registry_version: "2026-08-27.v1" },
    phase: { source_type: "publisher_current_milestone", source_field: "current_milestone", current_milestone: "Community Board hearing" },
    geography: { source_type: "affected_review_body_for", status: "resolved" },
    publisher: { source_type: "published_hearing", checked: true },
  },
  expected_next_stage: { stage_id: "ulurp_197c.borough_president_review", spine_phase_id: "borough_president", status: "known" },
  published_next_opportunity: { status: "none" },
  observed: { status: "no_observation", recommendations: [] },
  affected_actor_refs: [{ body_ref: "community-board:brooklyn-cb-11", role: "affected_community_board" }],
};

/**
 * Literal reproduction of site/app/land.mjs's landHearingRowHTML() markup,
 * with the new roleTxt fragment from landHearingRowRoleHTML() spliced in the
 * same position as the real function. The .py capture companion checks this
 * template's key substrings against the real source before rendering, so a
 * later edit to the real row markup fails the capture instead of silently
 * drifting from it.
 */
function hearingRowHTML({ title, borough, representing, whenLabel, modeTxt, venue, roleTxt, projectId }) {
  return `<div class="row land-hearing-row" data-i="0" data-project-id="${esc(projectId)}" tabindex="0" role="group">
    <p class="rtitle">${esc(title)}</p>
    <p class="rmeta"><span class="ragency">${esc(borough)}${representing ? ` · ${esc(representing)}` : ""}</span>${roleTxt ? ` · ${roleTxt}` : ""}
      · ${esc(t("land_hearings_card_when", { date: whenLabel }))}${modeTxt ? ` · ${esc(modeTxt)}` : ""}
      ${venue ? `<br>${esc(venue)}` : ""}
    </p>
    <div class="fcard-compact-actions"><a class="act" href="https://maps.example.test/phc06">${esc(t("land_action_attend_in_person"))}</a><a class="act" href="#land/${esc(projectId)}">${esc(t("land_hearings_open_project"))}</a></div>
  </div>`;
}

const PANEL_CASES = [
  {
    id: "decision_maker_with_prior_advisory",
    kind: "panel",
    assertion: "A1/A2: CPC's decision role renders in plain terms in the existing panel while the CB/BP advisory recommendations that preceded it stay visible as separate, still-labeled-advisory prior events.",
    summary: summarize("2025M0252"),
  },
  {
    id: "conditional_decision_maker",
    kind: "panel",
    assertion: "A1/A2: a conditional decision-maker (City Council under § 197-d) renders its own conditional plain-role sentence, distinct from an unconditional decider's.",
    summary: summarize("2026Q0210"),
  },
  {
    id: "advisory_only_never_final_decider",
    kind: "panel",
    assertion: "A2/negative rule: an advisory-only body renders \"Can recommend, but cannot decide\" — never described as the final decider.",
    summary: ADVISORY_ONLY_SUMMARY,
  },
  {
    id: "pre_application_certifier_no_vote",
    kind: "panel",
    assertion: "A3: a project in pre-application/certification (DCP, ELURP) renders \"Certifies the filing is complete, but does not decide it\" — it does not acquire a public-review vote.",
    summary: summarizeElurp("E1-2024Q0356"),
  },
  {
    id: "expedited_terminal_decision_maker",
    kind: "panel",
    assertion: "A4: an ELURP record terminal at CPC renders its own plain decision role without implying the ordinary-route Council stage this expedited profile omits.",
    summary: summarizeElurp("E2-2024Q0419"),
  },
  {
    id: "unresolved_no_role_honest_gap",
    kind: "panel",
    assertion: "A5/A6: a record whose current stage does not resolve renders no plain-role line at all — an honest gap, never an inferred disposition.",
    summary: summarizeElurp("E4-2026X0362"),
  },
];

const ROW_CASES = [
  {
    id: "hearing_row_decision_maker",
    kind: "row",
    assertion: "A1: the compressed upcoming-hearing row states \"Decides\" for a decision-maker body, reusing the same authority evidence as the panel.",
    row: { title: "165-05 Liberty Avenue Rezoning", borough: "Queens", representing: "City Planning Commission", whenLabel: "Wed, Oct 14", modeTxt: t("land_hearings_card_modes", { modes: t("land_hearings_mode_list_in_person") }), venue: "22 Reade Street, New York, NY", projectId: "2025M0252" },
    summary: summarize("2025M0252"),
  },
  {
    id: "hearing_row_advisory",
    kind: "row",
    assertion: "A2: the compressed row states \"Recommends, doesn't decide\" for an advisory body, never a decision verb.",
    row: { title: "Community Board Hearing", borough: "Brooklyn", representing: "Brooklyn Community Board 11", whenLabel: "Thu, Oct 8", modeTxt: t("land_hearings_card_modes", { modes: t("land_hearings_mode_list_in_person") }), venue: "123 Example Street, Brooklyn, NY", projectId: "PHC06-ADVISORY-ONLY" },
    summary: ADVISORY_ONLY_SUMMARY,
  },
  {
    id: "hearing_row_unresolved_no_role",
    kind: "row",
    assertion: "A5/A6: a row for a record with no resolved current role shows no role text at all — an honest gap, never a guessed one.",
    row: { title: "Referred Application", borough: "Bronx", representing: "City Council", whenLabel: "Fri, Oct 9", modeTxt: "", venue: "", projectId: "2026X0362" },
    summary: summarizeElurp("E4-2026X0362"),
  },
];

mkdirSync(OUT_DIR, { recursive: true });
const manifestCases = [];

for (const { id, kind, assertion, summary } of PANEL_CASES) {
  const html = landAuthoritySummaryHTML(summary, { t, escape: esc });
  const file = path.join(OUT_DIR, `${id}.html`);
  writeFileSync(file, html, "utf8");
  manifestCases.push({ id, kind, assertion, path: `/.phc06-capture-tmp/${id}.html`, project_id: summary.project_id || null });
}

for (const { id, kind, assertion, row, summary } of ROW_CASES) {
  const roleTxt = landHearingRowRoleHTML(summary, { t, escape: esc });
  const html = hearingRowHTML({ ...row, roleTxt });
  const file = path.join(OUT_DIR, `${id}.html`);
  writeFileSync(file, html, "utf8");
  manifestCases.push({ id, kind, assertion, path: `/.phc06-capture-tmp/${id}.html`, project_id: row.projectId });
}

process.stdout.write(JSON.stringify(manifestCases, null, 2));
